File: ra-idp/src/oauth2/usecases/refresh-tokens.ts
/**
 * Layer 3 — Application Logic
 *
 * リフレッシュトークンによる再発行。
 * ADR-004 のローテーション + ファミリー失効を実装する。
 */

import { hashToken, isAbsoluteExpired, isReplay, rotate } from '../domain/refresh-token'
import { OAuthError } from '../protocol/oauth-error'
import { evaluate } from '../../spec-bindings/policy/client-authorization'
import type { ClientRepository } from '../ports/client-repository'
import type { UserRepository } from '../../authentication/ports/user-repository'
import type { RefreshTokenStore } from '../ports/refresh-token-store'
import type { TokenIssuer } from '../ports/token-issuer'

export interface RefreshInput {
  client_id: string
  refresh_token: string
  /** DPoP / mTLS バインドがある場合の所有証明。 */
  proof_jkt?: string
  proof_x5t_s256?: string
}

export interface RefreshResult {
  access_token: string
  refresh_token: string
  token_type: 'Bearer' | 'DPoP'
  expires_in: number
  scope: string
}

export async function refreshTokenUseCase(
  deps: {
    clientRepo: ClientRepository
    userRepo: UserRepository
    refreshStore: RefreshTokenStore
    tokenIssuer: TokenIssuer
  },
  input: RefreshInput,
  emit: (event: { type: string; [k: string]: unknown }) => void,
  now: Date = new Date(),
): Promise<RefreshResult> {
  const client = await deps.clientRepo.findById(input.client_id)
  if (!client) {
    throw new OAuthError('invalid_client', 'クライアント認証に失敗しました')
  }

  const hash = hashToken(input.refresh_token)
  const record = await deps.refreshStore.findByHash(hash)
  if (!record) {
    throw new OAuthError('invalid_grant', 'リフレッシュトークンが無効です')
  }

  if (record.client_id !== client.client_id) {
    // 他人のリフレッシュトークンの使用 → 即座にファミリー失効
    await deps.refreshStore.revokeFamily(record.family_id)
    emit({
      type: 'RefreshTokenReuseDetected',
      occurredAt: now.toISOString(),
      familyId: record.family_id,
      tokenId: record.id,
      clientId: client.client_id,
    })
    throw new OAuthError('invalid_grant', 'リフレッシュトークンの所有者が一致しません')
  }

  if (isReplay(record)) {
    // ローテーション済み or 失効済み → ファミリー全失効 + アラート
    await deps.refreshStore.revokeFamily(record.family_id)
    emit({
      type: 'RefreshTokenReuseDetected',
      occurredAt: now.toISOString(),
      familyId: record.family_id,
      tokenId: record.id,
      clientId: client.client_id,
    })
    throw new OAuthError('invalid_grant', 'リフレッシュトークンはすでに使用されています')
  }

  if (isAbsoluteExpired(record, now)) {
    throw new OAuthError('invalid_grant', 'リフレッシュトークンが絶対期限切れです')
  }

  // ポリシー評価
  const decision = evaluate({
    subject: {
      type: 'Client',
      id: client.client_id,
      properties: { grantTypes: client.grant_types },
    },
    action: { name: 'token:grant_refresh' },
    resource: {
      type: 'RefreshToken',
      properties: {
        revoked: record.revoked,
        rotated: record.rotated,
        absoluteExpiresAt: record.absolute_expires_at,
        senderConstraint: record.sender_constraint ?? null,
      },
    },
    context: {
      proofOfPossession: record.sender_constraint
        ? {
            valid:
              (record.sender_constraint.type === 'dpop' &&
                record.sender_constraint.jkt === input.proof_jkt) ||
              (record.sender_constraint.type === 'mtls' &&
                record.sender_constraint['x5t#S256'] === input.proof_x5t_s256),
          }
        : null,
      now: now.toISOString(),
    },
  })
  if (decision.decision === 'Deny') {
    throw new OAuthError(
      'invalid_grant',
      `リフレッシュが拒否されました: ${decision.reasons?.join(', ')}`,
    )
  }

  // ADR-031: 無効化された user のトークン再発行はローテーション前に拒否する。
  // 状態を壊さないよう、refresh-store rotate より前に確認する。
  const user = await deps.userRepo.findBySub(record.sub)
  if (!user) throw new OAuthError('server_error', 'ユーザーが存在しません')
  if (user.disabled_at) {
    throw new OAuthError('invalid_grant', 'ユーザーは無効化されています')
  }

  // ローテーション: 新トークン発行 + 旧トークンを rotated にマーク
  const { token: newToken, record: newRecord } = rotate(record, now)
  const rotated = await deps.refreshStore.rotate(record.id, newRecord)
  if (!rotated) {
    // 並行ローテーション失敗 → ファミリー失効
    await deps.refreshStore.revokeFamily(record.family_id)
    throw new OAuthError('invalid_grant', '並行リフレッシュにより失効しました')
  }

  emit({
    type: 'RefreshTokenRotated',
    occurredAt: now.toISOString(),
    oldTokenId: record.id,
    newTokenId: newRecord.id,
    familyId: record.family_id,
  })

  const stored = record.sender_constraint
  const senderConstraint:
    | { type: 'dpop'; jkt: string }
    | { type: 'mtls'; 'x5t#S256': string }
    | null =
    stored && stored.type === 'dpop' && stored.jkt
      ? { type: 'dpop', jkt: stored.jkt }
      : stored && stored.type === 'mtls' && stored['x5t#S256']
        ? { type: 'mtls', 'x5t#S256': stored['x5t#S256'] }
        : null
  const { token: access_token } = await deps.tokenIssuer.signAccessToken({
    client,
    sub: record.sub,
    scopes: record.scopes,
    senderConstraint,
    authTime: Math.floor(now.getTime() / 1000),
  })

  return {
    access_token,
    refresh_token: newToken,
    token_type: senderConstraint ? 'DPoP' : 'Bearer',
    expires_in: deps.tokenIssuer.getAccessTokenTtlSeconds(),
    scope: record.scopes.join(' '),
  }
}

File: ra-idp-go/internal/oauth2/usecases/refresh_tokens.go
// リフレッシュトークンによる再発行。ADR-004 ローテーション + ファミリー失効。
package usecases

import (
	"context"
	"strings"
	"time"

	"ra-idp-go/internal/oauth2/domain"
	"ra-idp-go/internal/oauth2/ports"
	"ra-idp-go/internal/shared/spec"
	"ra-idp-go/internal/tenancy"
)

type RefreshInput struct {
	ClientID     string
	RefreshToken string
	ProofJKT     string
	ProofX5TS256 string
}

type RefreshResult struct {
	AccessToken  string
	RefreshToken string
	TokenType    string
	ExpiresIn    int
	Scope        string
}

type RefreshDeps struct {
	ClientRepo   ports.OAuth2ClientRepository
	UserRepo     ports.UserRepository
	RefreshStore ports.RefreshTokenStore
	TokenIssuer  ports.TokenIssuer
	Authorizer   ports.Authorizer
	Emit         func(spec.DomainEvent)
}

func RefreshTokens(ctx context.Context, deps RefreshDeps, in RefreshInput, now time.Time) (*RefreshResult, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	tenantID := tenancy.TenantID(ctx)
	client, err := deps.ClientRepo.FindByID(ctx, tenantID, in.ClientID)
	if err != nil {
		return nil, err
	}
	if client == nil {
		return nil, NewOAuthError("invalid_client", "クライアント認証に失敗")
	}
	hash := domain.HashRefreshToken(in.RefreshToken)
	record, err := deps.RefreshStore.FindByHash(ctx, hash)
	if err != nil {
		return nil, err
	}
	if record == nil {
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークンが無効")
	}
	if record.TenantID != tenantID {
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークンが無効")
	}
	if record.ClientID != client.ClientID {
		_ = deps.RefreshStore.RevokeFamily(ctx, record.FamilyID)
		emit(deps.Emit, &spec.RefreshTokenReuseDetected{At: now, TenantID: tenantID, FamilyID: record.FamilyID, TokenID: record.ID, ClientID: client.ClientID})
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークンの所有者が一致しません")
	}
	if domain.IsRefreshTokenReplay(record) {
		_ = deps.RefreshStore.RevokeFamily(ctx, record.FamilyID)
		emit(deps.Emit, &spec.RefreshTokenReuseDetected{At: now, TenantID: tenantID, FamilyID: record.FamilyID, TokenID: record.ID, ClientID: client.ClientID})
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークンはすでに使用されています")
	}
	if domain.IsRefreshTokenAbsoluteExpired(record, now) {
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークン絶対期限切れ")
	}
	user, err := deps.UserRepo.FindBySub(ctx, record.Sub)
	if err != nil {
		return nil, err
	}
	if user == nil {
		_ = deps.RefreshStore.RevokeFamily(ctx, record.FamilyID)
		return nil, NewOAuthError("invalid_grant", "ユーザーは利用できません")
	}
	if user.TenantID != tenantID {
		return nil, NewOAuthError("invalid_grant", "リフレッシュトークンが無効")
	}
	if !user.IsActive() {
		_ = deps.RefreshStore.RevokeFamily(ctx, record.FamilyID)
		return nil, NewOAuthError("invalid_grant", "ユーザーは無効化されています")
	}

	d, err := evaluateRefreshPolicy(ctx, deps.Authorizer, client, record, in, now)
	if err != nil {
		return nil, err
	}
	if !d.Permit {
		return nil, NewOAuthError("invalid_grant", "リフレッシュ拒否: "+strings.Join(d.Reasons, ", "))
	}

	newTok, err := domain.RotateRefreshToken(record, now)
	if err != nil {
		return nil, err
	}
	rotated, err := deps.RefreshStore.Rotate(ctx, record.ID, newTok.Record)
	if err != nil {
		return nil, err
	}
	if rotated == nil {
		_ = deps.RefreshStore.RevokeFamily(ctx, record.FamilyID)
		return nil, NewOAuthError("invalid_grant", "並行リフレッシュにより失効")
	}
	emit(deps.Emit, &spec.RefreshTokenRotated{At: now, TenantID: tenantID, OldTokenID: record.ID, NewTokenID: newTok.Record.ID, FamilyID: record.FamilyID})

	access, jti, err := deps.TokenIssuer.SignAccessToken(ctx, ports.AccessTokenInput{
		Client:           client,
		Sub:              record.Sub,
		Scopes:           record.Scopes,
		SenderConstraint: record.SenderConstraint,
		AuthTime:         now.Unix(),
	})
	if err != nil {
		return nil, err
	}
	emit(deps.Emit, &spec.AccessTokenIssued{At: now, TenantID: tenantID, JTI: jti, ClientID: client.ClientID, Sub: record.Sub, Scopes: record.Scopes, SenderConstraint: senderConstraintTag(record.SenderConstraint)})

	tokenType := "Bearer"
	if record.SenderConstraint != nil && record.SenderConstraint.Type == spec.SenderConstraintDPoP {
		tokenType = "DPoP"
	}
	return &RefreshResult{
		AccessToken:  access,
		RefreshToken: newTok.Token,
		TokenType:    tokenType,
		ExpiresIn:    deps.TokenIssuer.AccessTokenTTLSeconds(),
		Scope:        strings.Join(record.Scopes, " "),
	}, nil
}

func evaluateRefreshPolicy(
	ctx context.Context,
	authorizer ports.Authorizer,
	client *spec.OAuth2Client,
	record *spec.RefreshTokenRecord,
	in RefreshInput,
	now time.Time,
) (spec.AuthZResponse, error) {
	props := spec.AuthZResourceProps{
		Revoked:           record.Revoked,
		Rotated:           record.Rotated,
		AbsoluteExpiresAt: record.AbsoluteExpiresAt,
		SenderConstraint:  authZSenderConstraint(record.SenderConstraint),
	}
	var pop *spec.AuthZProofOfPossession
	if record.SenderConstraint != nil {
		valid := false
		switch record.SenderConstraint.Type {
		case spec.SenderConstraintDPoP:
			valid = record.SenderConstraint.JKT == in.ProofJKT
		case spec.SenderConstraintMTLS:
			valid = record.SenderConstraint.X5TS256 == in.ProofX5TS256
		}
		pop = &spec.AuthZProofOfPossession{Valid: valid}
	}
	req := spec.AuthZRequest{
		Subject:  spec.AuthZSubject{Type: "Client", ID: client.ClientID, Properties: spec.AuthZSubjectProps{GrantTypes: client.GrantTypes}},
		Action:   spec.ActionTokenGrantRefresh,
		Resource: spec.AuthZResource{Type: "RefreshToken", Properties: props},
		Context:  spec.AuthZContext{ProofOfPossession: pop, Now: now},
	}
	if authorizer == nil {
		return spec.Evaluate(req), nil
	}
	return authorizer.Authorize(ctx, req)
}

func authZSenderConstraint(sc *spec.SenderConstraint) *spec.AuthZSenderConstraint {
	if sc == nil {
		return nil
	}
	return &spec.AuthZSenderConstraint{Type: sc.Type}
}
